common/flusher/flusher.go

```diff
@@ -0,0 +1,244 @@
+	for i := 0; i < bufferCount-1; i++ { // -1 since flushItemsBuffer counts as the first free buffer
+		f.freeBufferChan <- make([]FlushItem[T], 0, bufferCapacity)
+	}
+}
+
+func (f *Flusher[T]) fullBufferChanHandler() {
+chanLoop:
+	for {
+		select {
+		case fullBuffer := <-f.fullBufferChan:
+			f.flushBuffer(fullBuffer)
+			freeBuffer := fullBuffer[:0]
+			f.freeBufferChan <- freeBuffer
+		case <-f.shutdownChan.Channel():
+			f.Stop()
+			break chanLoop
+		}
+	}
+}
+
+func (f *Flusher[T]) timerHandler() {
+TimerLoop:
+	for {
+		select {
+		case <-f.flushTimer.C:
+			f.Lock()
+			if len(f.flushItemsBuffer) > 0 {
+				f.clearBufferLocked()
+			}
+			f.Unlock()
+		case <-f.shutdownChan.Channel():
+			f.Stop()
+			break TimerLoop
+		}
+	}
+}
+
+func (f *Flusher[T]) getFreeBuffer() []FlushItem[T] {
+	var newFreeBuffer []FlushItem[T]
+	select {
+	case freeBuffer := <-f.freeBufferChan:
+		newFreeBuffer = freeBuffer
+	default:
+		newFreeBuffer = nil // set to nil to indicate no available flush buffer
+	}
+	return newFreeBuffer
+}
+
+func (f *Flusher[T]) clearBufferLocked() {
+	freeBuffer := f.getFreeBuffer()
+	fullBuffer := f.flushItemsBuffer
+	f.flushItemsBuffer = freeBuffer
+	f.fullBufferChan <- fullBuffer
+}
+
+func (f *Flusher[T]) insertAndClearBufferIfFullLocked(flushItem FlushItem[T]) {
+	if len(f.flushItemsBuffer) == 0 { // start timer if it's first item insertion
+		f.stopFlushTimer()
+		f.flushTimer.Reset(time.Duration(f.flushDuration))
+	}
+	f.flushItemsBuffer = append(f.flushItemsBuffer, flushItem)
+	if f.isBufferCloseToFullLocked(f.flushItemsBuffer) {
+		f.stopFlushTimer()
+		f.clearBufferLocked()
+	}
+}
+
+func (f *Flusher[T]) addItemToBeFlushed(item T) *future.FutureImpl[struct{}] {
+	f.Lock()
+	defer f.Unlock()
+	return f.addItemToBeFlushedLocked(item)
+}
+
+func (f *Flusher[T]) addItemToBeFlushedLocked(item T) *future.FutureImpl[struct{}] {
+	flushItem := FlushItem[T]{item, future.NewFuture[struct{}]()}
+	currFlushBuffer := f.flushItemsBuffer
+	if currFlushBuffer != nil { // nil check to make sure there is a usable flush buffer
+		f.insertAndClearBufferIfFullLocked(flushItem)
+	} else {
+		newFlushBuffer := f.getFreeBuffer()
+		if newFlushBuffer != nil { // nil check to make sure there is a usable flush buffer
+			f.flushItemsBuffer = newFlushBuffer
+			f.insertAndClearBufferIfFullLocked(flushItem)
+		} else {
+			err := serviceerror.NewUnavailable(fmt.Sprint("Flush buffer is currently full"))
+			flushItem.fut.Set(struct{}{}, err)
+		}
+	}
+
+	return flushItem.fut
+}
+
+func (f *Flusher[T]) flushBuffer(flushBuffer []FlushItem[T]) {
+	err := f.writer.Write(flushBuffer...)
+	if err != nil {
+		f.logger.Error("Flusher failed to write", tag.Error(err))
+	}
+	for _, flushItem := range flushBuffer {
+		flushItem.fut.Set(struct{}{}, err)
+	}
+}
+
+func (f *Flusher[T]) isBufferCloseToFullLocked(flushBuffer []FlushItem[T]) bool {
+	bufferCap := int(bufferCapacityGapPercent * float64(cap(flushBuffer)))
```

Comment:
Is this a fixed number? no need to calculate every time?

Reply:
Changed to a fixed number.